### ISSUE TITLE
How to set a threshold at inference time?

### ISSUE THREAD

**Issue by alevangel** [openvinotoolkit/anomalib#554]
I would like to set a threshold at test/inference time, so that I can tune the sensitivity of the model. 

How would be possible to do that?

**Comment by alexriedel1** [comment]
You can set `model.image_threshold = ... model.pixel_threshold = ... ` before inference

**Comment by alevangel** [comment]
@alexriedel1 this is not enough. Since if I set `model.image_threshold.value = ... model.pixel_threshold.value = ...  `, this will be overwritte by the function 
`trainer.predict(model=model, dataloaders=[dataloader])`, this function will also overwrite the model itself.

**Comment by alexriedel1** [comment]
Can you show me your testing / inferencing script? I'm pretty sure, calling `trainer.predict` does not affect the model

**Comment by alevangel** [comment]
This is the full implementation of my script

```
        model = get_model(config)
        callbacks = get_callbacks(config)

        trainer = Trainer(callbacks=callbacks, **config.trainer)

        # Set custom threshold
        model.adaptive_threshold = False
        model.pixel_threshold.value = torch.tensor(float(CUSTOM_NUMBER))
        model.image_threshold.value = torch.tensor(float(CUSTOM_NUMBER))

        transform_config = config.dataset.transform_config.val if "transform_config" in config.dataset.keys() else None
        dataset = InferenceDataset(
            my_args['input'], image_size=tuple(config.dataset.image_size), transform_config=transform_config
        )
        dataloader = DataLoader(dataset)
        trainer.predict(model=model, dataloaders=[dataloader])

```

At the end of this script my `model.image_threshold.value` will be the one tuned at training time.

**Comment by alexriedel1** [comment]
I cannot reproduce the issue with a Patchcore Model

```
    model.adaptive_threshold = False
    model.pixel_threshold.value = torch.tensor(float(1))
    model.image_threshold.value = torch.tensor(float(1))

    print(model.pixel_threshold.value, model.image_threshold.value)

    trainer.predict(model=model, dataloaders=[datamodule])

    print(model.pixel_threshold.value, model.image_threshold.value)
```
```
tensor(1.) tensor(1.)
LOCAL_RANK: 0 - CUDA_VISIBLE_DEVICES: [0]
Predicting DataLoader 0: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 3/3 [00:00<00:00, -4.90it/s]
tensor(1.) tensor(1.)
```

**Comment by alevangel** [comment]
Printing the threshold with    ` print(model.pixel_threshold.value, model.image_threshold.value)
` before and after the predict, what I get is this, I paste also the warnings I get

```
tensor(100.) tensor(100.)
\venv\lib\site-packages\torchmetrics\utilities\prints.py:36: UserWarning: Metric `ROC` will save all targets and predictions in buffer. For large datasets this may lead to large memory footprint.
  warnings.warn(*args, **kwargs)
\venv\lib\site-packages\torchmetrics\utilities\prints.py:36: UserWarning: Torchmetrics v0.9 introduced a new argument class property called `full_state_update` that has
                not been set for this class (MinMax). The property determines if `update` by
                default needs access to the full metric state. If this is not the case, significant speedups can be
                achieved and we recommend setting this to `False`.
                We provide an checking function
                `from torchmetrics.utilities import check_forward_no_full_state`
                that can be used to check if the `full_state_update=True` (old and potential slower behaviour,
                default for now) or if `full_state_update=False` can be used safely.
                
  warnings.warn(*args, **kwargs)
\venv\lib\site-packages\pytorch_lightning\trainer\connectors\data_connector.py:240: PossibleUserWarning: The dataloader, predict_dataloader 0, does not have many workers which may be a bottleneck. Consider increasing the value of the `num_workers` argument` (try 8 which is the number of cpus on this machine) in the `DataLoader` init to improve performance.
  rank_zero_warn(
Predicting DataLoader 0: 100%|██████████| 1/1 [00:00<00:00,  1.31it/s]
tensor(39.0126) tensor(25.4964)
```

**Comment by ashwinvaidya17** [comment]
@alevangel The thresholds are overwritten when `on_predict_start` is called from `LoadModelCallback`. One idea would be to manually call `torch.load` and pop `LoadModelCallback` from the callbacks list. Then updating the thresholds will work. This solution might work for your use case for now but we will try to come up with a better design.

Here is a snippet of the change to the `infer` function of `lightning_inference.py`

```python
def infer():
    """Run inference."""
    args = get_args()
    config = get_configurable_parameters(config_path=args.config)
    # This is commented as setting this adds LoadModelCallback to the callback list
    # config.trainer.resume_from_checkpoint = str(args.weights)
    config.visualization.show_images = args.show
    config.visualization.mode = args.visualization_mode
    if args.output:  # overwrite save path
        config.visualization.save_images = True
        config.visualization.image_save_path = args.output
    else:
        config.visualization.save_images = False

    model = get_model(config)
    model.load_state_dict(torch.load(args.weights)["state_dict"]) #  manually load weights
    callbacks = get_callbacks(config)

    trainer = Trainer(callbacks=callbacks, **config.trainer)

    transform_config = config.dataset.transform_config.val if "transform_config" in config.dataset.keys() else None
    dataset = InferenceDataset(
        args.input, image_size=tuple(config.dataset.image_size), transform_config=transform_config
    )
    dataloader = DataLoader(dataset)

    model.adaptive_threshold = False
    model.pixel_threshold.value = torch.tensor(float(1))
    model.image_threshold.value = torch.tensor(float(1))

    print(model.pixel_threshold.value, model.image_threshold.value)
    trainer.predict(model=model, dataloaders=[dataloader])
    print(model.pixel_threshold.value, model.image_threshold.value)
```

**Comment by alevangel** [comment]
@ashwinvaidya17 Thanks, this will work on avoid threshold overwriting. 
Owhever I had to add this to make it work:

```python
model_state_dict = torch.load(my_args['weights'], map_location=device)["state_dict"]
model_state_dict.pop("normalization_metrics.min", None)  # this is a key that doesn't make the model load the state
model_state_dict.pop("normalization_metrics.max", None)  # this is a key that doesn't make the model load the state
model.load_state_dict(model_state_dict)
```

But in this way the heatmap is not very normalized at all, even if the predicted mask is good.
